### ISSUE TITLE
Grafana Alerting: Alerts duplicate causing multiple alerts and resolution messages

### ISSUE THREAD

**Issue by ShaneGCareCru** [grafana/grafana#81198]
### What happened?

Hey everyone, I’m looking for some advice.

I get multiple notification for the same alert. I've tried recreating the alert with just the query and very basic settings but the same issue persisted. At one point I was getting 4 alerts but usually it's just 2. 

I have a loki alert that looks for the word “failed” in logs. It’s got a 30 minute window to look (the job runs every 30 min)
this is the query
count_over_time({instance="prod-backend", app="patientCache"} |= `Failed`[5m])
the history shows the duplication

![image](https://github.com/grafana/grafana/assets/103214468/8c85d9d4-9497-42c6-9f92-25034c671040)

### What did you expect to happen?

I expect that when you configure one alert, show preview and it says firing, that the alert would just appear once, not multiple times and also it's not grouped for those multiple, like others firing alerts would

### Did this work before?

I'm not sure

### How do we reproduce it?

1. Deploy grafana with helm chart on kubernetes (chart v7.0.3)
2. Create Loki alert
3. Multiple alerts

### Environment (with versions)?

Grafana: helm chart on k8s (7.0.3)
OS: N/A
Browser: N/A

### Grafana platform?

Kubernetes

### Datasource(s)?

loki

**Comment by tonypowa** [comment]
Hello @ShaneGCareCru 
Thank you for reporting this issue

n you provide the logs from Grafana around the time the alerts are fired and duplicated? These could contain useful information about what is happening internally.

**Comment by tgragnato** [comment]
I am also affected by this.

The bug is only reproducible in environments with HA enabled for unified_alerting.
Duplicate messages are sent again after a delay at least equal to ha_peer_timeout.

**Comment by ShaneGCareCru** [comment]
A reply on grafana's slack channel indicated that this was due to HA setup. The solutions were to turn of HA, or to have a dedicate non-HA grafana deployment for alerts. 

I opted for turning off HA, as that made sense for our use case. 

Ideally there would be a way to have alerting and HA. But I'm not sure if that's a feature request or a bug. 

Feel free to close this if this is a feature duplicate.

**Comment by KaplanSarah** [comment]
Please provide the specific steps to reproduce this issue.

**Comment by tgragnato** [comment]
Hello @KaplanSarah.
Please forgive the wait!

I think that given the response provided to @ShaneGCareCru, you already understand that the issue is a result of the **lack of byzantine fault tolerance in the ngalert scheduler**.

**_The behavior of the scheduler becomes unstable when instances in the cluster don't receive notifications at the same time_**.
At the end of the DedupStage discrepancies in the state means that alerts have resolved state but pipeline state firing.
At this point duplicated notifications gets sent with the timestamp of when they were created (or last updated).

The chance of this happening is increased during these practical situations:
- **rapid pod migrations in Kubernetes** (reorganizations of containers to handle peak loads)
- **rapid sequential restarts of vm processes** (ansible playbook executions with a small `serial:`)

**Comment by josh-freeman-tradeweb** [comment]
This is still broken

**Comment by ThuanPhuc27** [comment]
is there any solution for this

**Comment by josh-freeman-tradeweb** [comment]
> is there any solution for this

For me this was due to replicas, we had multiple pods of Grafana/Loki running. Once we setup HA to use a shared instance we now only get 1 alert, the grafana does show firing for multiple, but the actual alert to Slack/Email only comes through once.

**Comment by safi1315** [comment]
> is there any solution for this

Yes.  https://grafana.com/docs/grafana/latest/alerting/set-up/configure-high-availability/.

If your grafana consists of multiple nodes, you need to follow this. This worked for me

**Comment by safi1315** [comment]
> licas, we had multiple pods of Grafana/Loki running. Once we setup HA to use a shared instance we now only get 1 alert, the grafana does show firing for multiple, but the actual alert to Slack/Email only comes throu

Yup this worked for me too. https://grafana.com/docs/grafana/latest/alerting/set-up/configure-high-availability/

**Comment by tgragnato** [comment]
> > licas, we had multiple pods of Grafana/Loki running. Once we setup HA to use a shared instance we now only get 1 alert, the grafana does show firing for multiple, but the actual alert to Slack/Email only comes throu
> 
> Yup this worked for me too. https://grafana.com/docs/grafana/latest/alerting/set-up/configure-high-availability/

The bug exists because of how memberlist behaves in HA setups. Following the HA configuration docs isn't a fix: it's actually the environment where the bug manifests.
